File: 03-creational-patterns/02-abstract-factory.go
package main

import "fmt"

// Abstract Products
type Button interface {
	Paint()
}

type Checkbox interface {
	Paint()
}

// Concrete Products - Windows
type WindowsButton struct{}

func (WindowsButton) Paint() { fmt.Println("Rendering Windows Button") }

type WindowsCheckbox struct{}

func (WindowsCheckbox) Paint() { fmt.Println("Rendering Windows Checkbox") }

// Concrete Products - MacOS
type MacButton struct{}

func (MacButton) Paint() { fmt.Println("Rendering Mac Button") }

type MacCheckbox struct{}

func (MacCheckbox) Paint() { fmt.Println("Rendering Mac Checkbox") }

// Abstract Factory
type GUIFactory interface {
	CreateButton() Button
	CreateCheckbox() Checkbox
}

// Concrete Factories
type WindowsFactory struct{} //implements GUIFactory

func (WindowsFactory) CreateButton() Button {
	return WindowsButton{}
}
func (WindowsFactory) CreateCheckbox() Checkbox {
	return WindowsCheckbox{}
}

type MacFactory struct{} //implements GUIFactory

func (MacFactory) CreateButton() Button {
	return MacButton{}
}
func (MacFactory) CreateCheckbox() Checkbox {
	return MacCheckbox{}
}

// Client
func renderUI(factory GUIFactory) {
	btn := factory.CreateButton()
	chk := factory.CreateCheckbox()
	btn.Paint()
	chk.Paint()
}

func main() {
	var factory GUIFactory

	// simulate OS
	os := "mac"
	if os == "windows" {
		factory = WindowsFactory{}
	} else {
		factory = MacFactory{}
	}

	renderUI(factory)
}

File: 03-creational-patterns/02-abstract-factory/02-abstract-factory.py
from abc import ABC, abstractmethod

# Abstract Products
class Button(ABC):
    @abstractmethod
    def paint(self): pass

class Checkbox(ABC):
    @abstractmethod
    def paint(self): pass

# Concrete Products - Windows
class WindowsButton(Button):
    def paint(self):
        print("Rendering Windows Button")

class WindowsCheckbox(Checkbox):
    def paint(self):
        print("Rendering Windows Checkbox")

# Concrete Products - MacOS
class MacButton(Button):
    def paint(self):
        print("Rendering Mac Button")

class MacCheckbox(Checkbox):
    def paint(self):
        print("Rendering Mac Checkbox")

# Abstract Factory
class GUIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button: pass

    @abstractmethod
    def create_checkbox(self) -> Checkbox: pass

# Concrete Factories
class WindowsFactory(GUIFactory):
    def create_button(self) -> Button:
        return WindowsButton()
    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()

class MacFactory(GUIFactory):
    def create_button(self) -> Button:
        return MacButton()
    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()

# Client
def render_ui(factory: GUIFactory):
    btn = factory.create_button()
    chk = factory.create_checkbox()
    btn.paint()
    chk.paint()

if __name__ == "__main__":
    os = "mac"
    factory = MacFactory() if os == "mac" else WindowsFactory()
    render_ui(factory)
